components/playlist/playlist_download_request_manager.h

```diff
@@ -0,0 +1,89 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_PLAYLIST_PLAYLIST_DOWNLOAD_REQUEST_MANAGER_H_
+#define BRAVE_COMPONENTS_PLAYLIST_PLAYLIST_DOWNLOAD_REQUEST_MANAGER_H_
+
+#include <list>
+#include <memory>
+#include <string>
+
+#include "base/memory/weak_ptr.h"
+#include "base/scoped_observer.h"
+#include "brave/components/playlist/playlist_types.h"
+#include "brave/components/playlist/playlist_youtubedown_component_manager.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_contents_observer.h"
+
+namespace base {
+class Value;
+}  // namespace base
+
+namespace content {
+class BrowserContext;
+}  // namespace content
+
+namespace playlist {
+
+struct CreatePlaylistParams;
+
+class PlaylistDownloadRequestManager :
+    public PlaylistYoutubeDownComponentManager::Observer,
+    public content::WebContentsObserver {
+ public:
+  class Delegate {
+   public:
+    virtual void OnPlaylistCreationParamsReady(
+        const CreatePlaylistParams& params) = 0;
+  };
+
+  static void SetPlaylistJavaScriptWorldId(const int32_t id);
+
+  PlaylistDownloadRequestManager(content::BrowserContext* context,
+                                 Delegate* delegate,
+                                 PlaylistYoutubeDownComponentManager* manager);
+  ~PlaylistDownloadRequestManager() override;
+  PlaylistDownloadRequestManager(
+      const PlaylistDownloadRequestManager&) = delete;
+  PlaylistDownloadRequestManager& operator=(
+      const PlaylistDownloadRequestManager&) = delete;
+
+  // Delegate will get called with generated param.
+  void GeneratePlaylistCreateParamsForYoutubeURL(const std::string& url);
+
+ private:
+  // PlaylistYoutubeDownComponentManager::Observer overrides:
+  void OnYoutubeDownScriptReady(const std::string& youtubedown_script) override;
+
+  // content::WebContentsObserver overrides:
+  void DidFinishLoad(content::RenderFrameHost* render_frame_host,
+                     const GURL& validated_url) override;
+
+  // Will get date from youtube for downloading media files of |url| by
+  // injecting youtubedown.js.
+  void FetchYoutubeDownData(const std::string& url) const;
+  void OnGetYoutubeDownData(base::Value value);
+  bool ReadyToRunYoutubeDownJS() const;
+  void CreateWebContents();
+  void FetchAllPendingYoutubeURLs();
+
+  std::list<std::string> pending_youtube_urls_;
```

Comment:
I would add a comment describing this in some more detail

Reply:
Added.